MoreLinq/Experimental/SelectAsync.cs

```diff
@@ -0,0 +1,519 @@
+#region License and Terms
+// MoreLINQ - Extensions to LINQ to Objects
+// Copyright (c) 2016 Atif Aziz. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#endregion
+
+#if !NO_ASYNC
+
+namespace MoreLinq.Experimental
+{
+    using System;
+    using System.Collections;
+    using System.Collections.Concurrent;
+    using System.Collections.Generic;
+    using System.Diagnostics;
+    using System.Linq;
+    using System.Runtime.ExceptionServices;
+    using System.Threading;
+    using System.Threading.Tasks;
+
+    /// <summary>
+    /// Represents options for an asynchronous projection operation.
+    /// </summary>
+
+    public sealed class SelectAsyncOptions
+    {
+        /// <summary>
+        /// The default options an asynchronous projection operation.
+        /// </summary>
+
+        public static readonly SelectAsyncOptions Default =
+            new SelectAsyncOptions(null /* = unbounded concurrency */,
+                                   TaskScheduler.Default,
+                                   preserveOrder: false);
+
+        /// <summary>
+        /// Gets a positive (non-zero) integer that specifies the maximum
+        /// projections to run concurrenctly or <c>null</c> to mean unlimited
+        /// concurrency.
+        /// </summary>
+
+        public int? MaxConcurrency { get; }
```

Comment:
Why not non-nullable int? 0 is not allowed anyway.

Reply:
I wanted to start with a nullable integer and see how it looks and feel. I agree that it's not any clearer to anyone what `null` would mean here besides undefined and that undefined means unbounded. Both 0 and `null` would need that to be spelled out in the documentation.

Reply:
> I agree that it's not any clearer to anyone what `null` would mean here

On second thoughts, `MaxConcurrency(0)` reads just worse than `MaxConcurrency(null)`. With the former, it's an invalid value overloaded to be [magical](https://en.wikipedia.org/wiki/Magic_number_(programming)) and one can only take a wild guess without looking up the documentation. With `null`, an undefined maximum, could really only mean one thing for concurrency: no maximum, therefore unbounded. The null here means an _optional maximum_ and type of `int?` communicates that better. I'm inclined to stay with this design.

Reply:
Given there is already a `AsSequential`, maybe there should be `MaxConcurrency(int)` to set a limit, and `UnboundedConcurrency` to set it to null.